### PR TITLE
feat: add hanoi

### PR DESCRIPTION
**Pull request by toshuyu**
add a hanoi tower implementation

### CODE REVIEW

**Comment by toshuyu** [comment]
Here's RSD review on behalf of masnec. (review count: 1)
 I have reviewed the provided Pull Request and the new `script.js` file. Here are my comments and suggestions:

1. Unit tests: Since the original diff doesn't contain unit tests, I have written some basic Jest unit tests for the `TowersOfHanoi` class. You can find the tests in the `towers-of-hanoi.test.js` file below.

```javascript
const TowersOfHanoi = require('./script');

describe('TowersOfHanoi', () => {
  let hanoi;

  beforeEach(() => {
    hanoi = new TowersOfHanoi(3);
  });

  test('initial state', () => {
    expect(hanoi.towers).toEqual([[3, 2, 1], [], []]);
    expect(hanoi.moves).toBe(0);
  });

  test('moveDisk', () => {
    hanoi.moveDisk(0, 1);
    expect(hanoi.towers).toEqual([[2, 1], [3], []]);
    expect(hanoi.moves).toBe(1);
  });

  test('solve', () => {
    hanoi.solve();
    expect(hanoi.towers).toEqual([[], [], [3, 2, 1]]);
    expect(hanoi.moves).toBe(7);
  });
});
```

2. Code style and best practices:

- The code is well-structured and easy to read.
- The class `TowersOfHanoi` follows the Single Responsibility Principle, as it handles both the game state and the logic of solving the Towers of Hanoi problem.
- The constructor initializes the game state, and the methods `moveDisk`, `solve`, `solveRecursive`, and `printState` handle specific actions or queries related to the game.
- The code is consistent with the provided example usage, demonstrating how to use the `TowersOfHanoi` class.

3. Maintainability:

- The code is easy to maintain, as it is well-organized and follows common object-oriented programming principles.
- Adding new features or modifying existing functionality should be straightforward.

4. Performance:

- The performance of the code is acceptable for the given problem, as the Towers of Hanoi problem is not computationally intensive.
- The recursive `solveRecursive` method may cause a stack overflow for a large number of disks, but this is inherent to the problem itself.

5. Security:

- Security is not a concern for this code, as it is a simple console application without external inputs or interactions.

In conclusion, the code is well-written, clean, and follows best practices. I recommend merging the Pull Request after addressing the unit test requirement.

**Comment by toshuyu** [comment]
Here's RSD review on behalf of derwin162. (review count: 1)
 I have reviewed the provided Pull Request and checked it against the outlined tasks and software engineering best practices. Here are my findings:

1. Typos: I did not find any typos in the code.
2. Variable naming: The variable naming is meaningful and follows best practices. For example, `numDisks`, `towers`, `source`, `destination`, `auxiliary`, and `moves` are all descriptive and easy to understand.

Additional comments and suggestions:

1. Comments: While the code is quite clear and easy to follow, adding a brief comment at the beginning of the `solveRecursive` method explaining the base case and recursive steps would be helpful for someone unfamiliar with the Tower of Hanoi problem.
2. Constants: Consider defining a constant for the number of towers (currently 3) to make it easier to modify the code if needed.
3. Encapsulation: The `printState` method is currently public, but it might be more appropriate to make it a private method and provide a separate `getState` method that returns the current state of the towers as an array or object. This would allow for more flexibility in how the state is displayed or used.
4. Testing: It would be beneficial to add unit tests for this class to ensure that it functions correctly for various inputs and edge cases.

Overall, the code looks clean, well-organized, and follows best practices. Good job!

**Comment by toshuyu** [comment]
test doublecheck for masnec

**Comment by toshuyu** [comment]
test done for derwin162

**Comment by toshuyu** [comment]
Hi @derwin162,
RSD has helped you finished review. Please do manual review now, thanks.

**Comment by toshuyu** [comment]
test doublecheck for masnec

**Comment by toshuyu** [comment]
Here's RSD review on behalf of masnec. (review count: 3)
 I have reviewed the provided Pull Request, and here are my comments and observations:

1. The code adheres to the Single Responsibility Principle by separating the game logic and test code into different files: `script.js` and `test.js`.
2. The `TowersOfHanoi` class is well-defined with proper encapsulation and follows the object-oriented programming paradigm.
3. The class constructor initializes the game state, and the methods `moveDisk`, `solve`, `solveRecursive`, and `printState` handle the game logic.
4. The test suite in `test.js` covers the initial state, `moveDisk`, and `solve` methods, ensuring that the code functions as expected.
5. The code is clean, easy to read, and maintainable.

However, I would like to suggest a few improvements:

1. Add comments to the `solveRecursive` method to make it clearer how the Tower of Hanoi algorithm works.
2. In the `moveDisk` method, add error handling for invalid tower indices.
3. In the test suite, add more test cases to cover edge cases, such as moving a disk from an empty tower or moving a disk larger than the target tower's top disk.
4. In the `solve` method, consider adding a check to ensure that the number of disks is a positive integer.
5. In the `solveRecursive` method, add a base case for `numDisks` equal to 0 to avoid unnecessary recursive calls.

Overall, the code is well-written and follows best practices. With the suggested improvements, the code will be even more robust and maintainable.

**Comment by toshuyu** [comment]
test done for masnec

**Comment by toshuyu** [comment]
Hi @masnec,
RSD has helped you finished review. Please do manual review now, thanks.
